### PR TITLE
Move Security to use auto-managed system indices

### PR DESCRIPTION
**Pull request by pugnascotia**
Backport of #67114. Part of #61656.

Change the Security plugin so that its system indices are managed automatically
by the system indices infrastructure.

Also add an `origin` field to `CreateIndexRequest` and `UpdateSettingsRequest`.

### CODE REVIEW

**Comment by mark-vieira** [comment]
@elasticmachine update branch

**Comment by pugnascotia** [comment]
@jaymode / @albertzaharovits there was a test failure in this backport:

https://gradle-enterprise.elastic.co/s/2vx2uqtnhvfc6/tests/:x-pack:qa:rolling-upgrade:v6.8.14%23oneThirdUpgradedTest/org.elasticsearch.upgrades.UpgradeClusterClientYamlTestSuiteIT/test%20%7Bp0=mixed_cluster%2F120_api_key_auth%2FTest%20API%20key%20authentication%20will%20work%20in%20a%20mixed%20cluster%7D#1

I think it's legitimate but it's intermittent - looks like in an upgrade scenario, a test tried to create an API key, but it couldn't because the mappings on `.security` weren't updated yet from the 6.8 mappings, so they lacked a mapping for `realm_type`. This situation would have worked before, because although strict mapping was in effect, the `SecurityIndexManager` would have upgraded the mappings before indexing a document.

I could reinstate that behaviour in `SecurityIndexManager`, but then:

   - It would be doing something different to `SystemIndexManager`
   - What mappings should it apply in mixed version cluster?

Or, I could change `ApiKeyService` to only index the `realm_type` field if the `.security` index has a mapping for `realm_type`. I'm not sure what I think about that.

Thoughts appreciated.

**Comment by jaymode** [comment]
In these situations, I would prefer to be able to support these features in a mixed version cluster if the changes to the index can be applied (the index change doesn't require a feature only available in a new version). For the purposes of this PR, I'd put the behavior into `SecurityIndexManager`.

As a later TODO I think we could generalize this behavior into a special system index class/client that would enable non-breaking updates during rolling upgrades independent of the master version.

**Comment by albertzaharovits** [comment]
> For the purposes of this PR, I'd put the behavior into SecurityIndexManager.

I agree.
Looks like in 7.x, in a mixed scenario from 6.8, creating an API key on the new node and then another new APi key on some old node my turn up to be problematic. For this reason, I would maintain the existing behavior. I'll take it to investigate and raise an issue.

File: x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java [hunk 1]
```diff
@@ -219,6 +226,19 @@ private boolean checkIndexAvailable(ClusterState state) {
         }
     }
 
+    private boolean checkIndexMappingUpToDate(ClusterState clusterState) {
+        return checkIndexMappingVersionMatches(clusterState, Version.CURRENT::equals);
```
Comment (jaymode) [inline x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java:230]: Shouldn't this be `onOrAfter`?

Comment (pugnascotia) [reply x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/support/SecurityIndexManager.java]: Ah, good catch 👍

File: server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java [hunk 2]
```diff
@@ -66,9 +70,25 @@ protected void masterOperation(final CreateIndexRequest request, final ClusterSt
...
-            : buildUpdateRequest(request, cause, indexName);
+        final boolean isSystemIndex = descriptor != null && descriptor.isAutomaticallyManaged();
+
+        final CreateIndexClusterStateUpdateRequest updateRequest;
+
+        // Requests that a cluster generates itself are permitted to create a system index with
+        // different mappings, settings etc. This is so that rolling upgrade scenarios still work.
+        // We check this via the request's origin. Eventually, `SystemIndexManager` will reconfigure
+        // the index to the latest settings.
+        if (isSystemIndex && Strings.isNullOrEmpty(request.origin())) {
```
Comment (jaymode) [inline server/src/main/java/org/elasticsearch/action/admin/indices/create/TransportCreateIndexAction.java:81]: This is pretty minor but `request.origin()` should never be null based on my reading of the code. Is this correct? If it should never be null, then I believe we should just call `request.origin().isEmpty()` here and the other two places.

**Review by jaymode — APPROVED** [review]
Left one comment, but otherwise LGTM

**Pull request merged**